packages/babel-plugin-transform-react-jsx/src/index.js

```diff
@@ -1,18 +1,67 @@
-/* eslint-disable-next-line @babel/development/plugin-name */
-import transformClassic from "./transform-classic";
-/* eslint-disable-next-line @babel/development/plugin-name */
-import transformAutomatic from "./transform-automatic";
+import jsx from "@babel/plugin-syntax-jsx";
+import { helper } from "@babel/helper-builder-react-jsx-experimental";
 import { declare } from "@babel/helper-plugin-utils";
+import { types as t } from "@babel/core";
 
 export default declare((api, options) => {
   const { runtime = "classic" } = options;
+  const PURE_ANNOTATION = options.pure;
 
-  // we throw a warning in helper-builder-react-jsx-experimental if runtime
-  // is neither automatic or classic because we will remove this file
-  // in v8.0.0
-  if (runtime === "classic") {
-    return transformClassic(api, options);
-  } else {
-    return transformAutomatic(api, options);
+  if (
+    options.useSpread !== undefined &&
+    runtime === "classic" &&
+    typeof options.useSpread !== "boolean"
+  ) {
+    throw new Error(
+      "transform-react-jsx currently only accepts a boolean option for " +
+        "useSpread (defaults to false)",
+    );
   }
+
+  const visitor = helper(api, {
+    pre(state) {
+      const tagName = state.tagName;
+      const args = state.args;
+      if (t.react.isCompatTag(tagName)) {
+        args.push(t.stringLiteral(tagName));
+      } else {
+        args.push(state.tagExpr);
+      }
+    },
+
+    post(state, pass) {
+      if (pass.get("@babel/plugin-react-jsx/runtime") === "classic") {
+        state.createElementCallee = pass.get(
+          "@babel/plugin-react-jsx/createElementIdentifier",
+        )();
+
+        state.pure =
+          PURE_ANNOTATION ?? !pass.get("@babel/plugin-react-jsx/pragmaSet");
+      } else {
+        state.jsxCallee = pass.get("@babel/plugin-react-jsx/jsxIdentifier")();
+        state.jsxStaticCallee = pass.get(
+          "@babel/plugin-react-jsx/jsxStaticIdentifier",
+        )();
+        state.createElementCallee = pass.get(
+          "@babel/plugin-react-jsx/createElementIdentifier",
+        )();
+
+        state.pure =
+          PURE_ANNOTATION ??
+          !pass.get("@babel/plugin-react-jsx/importSourceSet");
+      }
+    },
+
+    ...options,
+    development: false,
+    runtime,
+    useSpread:
+      "useSpread" in options ? options.useSpread : runtime !== "classic",
```

Comment:
Can we throw if useSpread is set with the automatic runtime?

Reply:
This is somewhat tricky because the `runtime` option doesn't match the actual runtime that will be used when transforming.

What should happen for:
```
// config
const options = { runtime: 'classic', useSpread: false }

// ...
/** @jsxRuntime automatic */

<div {...props} key="foo" />
``` 

If we want to keep using extends helper etc here (which seems rather logical) then this would also imply that we should implement extends helper support for `jsx`, `jsxs` and `jsxDEV` outputs, right?

Reply:
> If we want to keep using extends helper etc here (which seems rather logical)

No, this should also throw.

I think that we can throw if:
- `runtime: "automatic"` and `useSpread: false` are set in the options (when instantiating the plugin)
- `useSpread: false` is explicitly set in the options and there is a `/** @jsxRuntime automatic */` comment (when traversing)

Reply:
> runtime: "automatic" and useSpread: false are set in the options (when instantiating the plugin)

Technically this was previously allowed, even if `useSpread` has been silently ignored in such a case. I'm OK with adding an error in a situation like this - I just want to ensure myself that you are as well.

> useSpread: false is explicitly set in the options and there is a /** @jsxRuntime automatic */ comment (when traversing) <-- In this case it's maybe also ok to just warn and ignore useSpread: false, depending on what would be a breaking change.

Previously the `/** @jsxRuntime automatic */` was completely ignored if the runtime was set to `"classic"` at the plugin level. Technically one could have it in a file but it would have no effect. 

---

So what you are proposing is that whenever we use the automatic runtime we should ignore options like `useSpread` regardless of if the runtime has been chosen at the plugin or a file level? So to make those options only applicable when one uses the classic runtime through options (or by just using the default which is still the classic) **and** not changing the runtime at the file level (so when it stays to be classic). Did I get the intention right?

Reply:
> So what you are proposing is that whenever we use the automatic runtime we should ignore options like useSpread regardless of if the runtime has been chosen at the plugin or a file level? So to make those options only applicable when one uses the classic runtime through options (or by just using the default which is still the classic) and not changing the runtime at the file level (so when it stays to be classic). Did I get the intention right?

Yes, that's exactly what I'd like to see.